rfd/0080-piv-private-key.md

```diff
@@ -0,0 +1,316 @@
+
+While this strategy does prevent the user from needing to login twice, it presents a possible workaround. If a Teleport Client stores the initial RSA private key and certificates, then they can be used by anyone to reissue hardware key certifificates. To mitigate this risk, the initial certificates will have a TTL of 1 minute and will only be authorized for a single reissue request. This flow is similar to per-session MFA, which also issues one-time-use certificates with a 1 minute TTL.
+
+To avoid situtations where the user has to login twice, we will go with the second option.
+
+#### Hardware private key login
+
+As mentioned above, hardware key login will start with the normal login flow, followed by a call to `GetPrivateKeyPolicy`. If the result is `PRIVATE_KEY_POLICY_HARDWARE_KEY` or `PRIVATE_KEY_POLICY_HARDWARE_KEY_TOUCH`, then the Teleport client will:
+ 1. Find a hardware key on the user's device
+ 2. Generate a new private key on the hardware key (with Touch policy "cached" if required)
+ 3. Use the hardware private key to get a new set of signed certificates from the Teleport Auth Server
+ 4. Upsert Attestation data with `AttestHardwarePrivateKey` so that future Teleport requests can verify that the certificates being used chain back to a trusted hardware key CA 
+
+The resulting certificates from will only be operable if:
+ - The hardware key is connected during the operation
+ - The hardware private key can still be found
+ - The hardware private keys' Touch challenges are passed (if applicable)
+
+#### PIV slot overwrite/reuse logic
+
+For PIV hardware keys, we will use slot `9a`, which is reserved for authentication. Before generating a new private key, Teleport clients will check if PIV slot `9a` is already in use by a Teleport client. If it is, then the client will reuse the existing private key instead of generating a new one. Otherwise, it will generate a new private key.
+
+#### Private key interface
+
+Currently, Teleport clients store a PEM encoded private key (`~/.tsh/keys/proxy/user`) for a login session. This PEM encoded private key is then unmarshalled, transformed, and parsed as needed during a client request.
+
+With a hardware private key, we only have access to a raw `crypto.PrivateKey`, and do not have sufficient information about the key to transform it into a `*rsa.PrivateKey` and marshal it into PKCS1 format. Instead, we need to alter Teleport clients to use `cyrpto.PrivateKey` by default. This will require altering the key interface (`lib/client/interfaces.go`) and its usage across `lib/client` and other relevant locations. `lib/utils/native` will also be updated to return `*rsa.PrivateKey` instead of its PEM encoded private and public keys.
```

Comment:
This sounds like a good change. Could we remove the baked assumptions that keys are RSA, while we are at it?  (Or at least make the assumption the least possible?)

It would be nice to switch to ed25519 one day. :)

Reply:
Yes, already done in `https://github.com/gravitational/teleport/pull/15334` since it looks like Teleport Connect was using ECDSA keys sometimes.

Reply:
Nice^2!